### PR TITLE
tree2: hide SharedTreeFactory and TreeFactory

### PR DESCRIPTION
**Pull request by DLehenbauer**
SharedTreeFactory and TreeFactory transitively expose non-public types via their configurable options.

Rather than surface these types, I updated code outside the tree package to use 'SharedTree.getFactory()' to create a SharedTree.

SharedTree.getFactory() must be publicly exposed for AzureClient.

### CODE REVIEW

File: packages/dds/tree/api-report/tree.api.md [hunk 1]
```diff
@@ -1577,7 +1577,7 @@ export interface SequenceFieldEditBuilder {
     move(sourceIndex: number, count: number, destIndex: number): void;
 }
 
-// @internal
+// @beta
 export class SharedTree implements ITree {
```
Comment (DLehenbauer) [inline packages/dds/tree/api-report/tree.api.md:1581]: This is our faux 'SharedTree' class we expose to comply with the AzureClient pattern, which requires a class with a specific ctor signature.

File: packages/dds/tree/api-report/tree.api.md [hunk 2]
```diff
@@ -1586,7 +1586,7 @@ export class SharedTree implements ITree {
     // (undocumented)
     getAttachSummary(fullTree?: boolean | undefined, trackState?: boolean | undefined, telemetryContext?: ITelemetryContext | undefined): ISummaryTreeWithStats;
     // (undocumented)
-    static getFactory(): TreeFactory;
+    static getFactory(): IChannelFactory;
```
Comment (DLehenbauer) [inline packages/dds/tree/api-report/tree.api.md:1589]: Avoid exposing 'TreeFactory'.  AzureClient blindly casts the IChannel to the type of the class from which it received its factory (via .getFactory()).

Note that despite it being the prevailing pattern in our code, calling 'runtime.create(..)' or 'IChannelFactory.create(..)' directly is not publicly supported by the SDK.

File: packages/tools/devtools/devtools-core/src/test/DefaultVisualizers.test.ts [hunk 3]
```diff
@@ -394,10 +394,10 @@ describe("DefaultVisualizers unit tests", () => {
 	});
 
 	it("SharedTree", async () => {
-		const factory = new TreeFactory({});
+		const factory = SharedTree.getFactory();
 		const builder = new SchemaFactory("DefaultVisualizer_SharedTree_Test");
 
-		const sharedTree = factory.create(new MockFluidDataStoreRuntime(), "test");
+		const sharedTree = factory.create(new MockFluidDataStoreRuntime(), "test") as ITree;
```
Comment (alexvy86) [inline packages/tools/devtools/devtools-core/src/test/DefaultVisualizers.test.ts:400]: Is this how it works for other DDS factories today? Consumers are forced to cast?

Comment (DLehenbauer) [reply packages/tools/devtools/devtools-core/src/test/DefaultVisualizers.test.ts:400]: Yes, if you're invoking the result of 'getFactory()' directly.

For example, you can see that SharedMap.getFactory() also returns an untyped IChannelFactory:
```ts
	public static getFactory(): IChannelFactory {
		return new MapFactory();
	}
```

No, if you use AzureClient / ContainerSchema, in which case no cast is necessary:

```ts
		it("preserves 'SharedTree' type", async () => {
			const { container } = await client.createContainer({
				initialObjects: {
					tree: SharedTree,
				},
			});

			// Ensure that the 'tree' API is accessible without casting or suppressing lint rules:
			const tree: ITree = container.initialObjects.tree;
```

**Review by taylorsw04 — APPROVED** [review]

**Review by taylorsw04 — APPROVED** [review]

**Comment by CraigMacomber** [comment]
I think there might be value in allowing users of our package to opt into addition validation and asserts, which is all these options you are hiding do, especially if they are reporting bugs to us in the beta period. We may wany to consider keeping some way to do that.

**Comment by DLehenbauer** [comment]
@CraigMacomber - Our strategy is to trim aggressively, but also export the untrimmed .d.ts file via '/internal' as a "universal escape hatch".  If we need to help a customer debug an issue, we can point them at TreeFactory at this path:
```ts
import { TreeFactory } from '@fluidframework/tree/internal'
```

**Pull request merged**